Tests/AsyncHTTPClientTests/HTTPClientTests.swift

```diff
@@ -380,9 +384,12 @@ class HTTPClientTests: XCTestCase {
     }
 
     func testProxyTLS() throws {
+        XCTFail("Disabled test as it crashes");
```

Comment:
Looking into it, for some reason HTTPBin is not setup correctly

Reply:
ok, so this test actually catches one very specific behaviour that is broken by this PR. TLS proxy works as following: we create a pipeline (without SSL handlers), then send `CONNECT` in plaintext, remove proxy handler and add SSL handlers. In this PR we enable TLS right from the start, meaning that `CONNECT` is sent encrypted. cc @weissi @Lukasa is it possible to bypass TLS or delay its initialization?

Reply:
@artemredkin yes, just add the SSL handler at exactly the point from which you want encryption from.

Reply:
here's an example for STARTTLS in IMAP: We add the SSL handler after sending the `STARTTLS` command: https://github.com/apple/swift-nio-examples/blob/master/NIOSMTP/NIOSMTP/SendEmailHandler.swift#L132-L147

Reply:
@weissi this is what we do right now, question is what to do about NIOTS? SSL is handled by TS...

Reply:
@artemredkin sadly Network.framework doesn't support this. The only way to get this to work is to tell Network.framework (and therefore NIOTS) to _not_ use TLS and then add the the SSL handler from NIO SSL :). NIO SSL's SSL handler is compatible with all implementations.

If you check the NIOSMTP example, this time at https://github.com/apple/swift-nio-examples/blob/4bdfd14f416df123261714cdddbc8d812fa10acb/NIOSMTP/NIOSMTP/ViewController.swift#L132-L159 you'll also see that's exactly what we do there

Reply:
For regular TLS, we bundle `NIOTSConnectionBootstrap` with its regular `NIOTSClientTLSProvider`

```
            bootstrap = NIOClientTCPBootstrap(NIOTSConnectionBootstrap(group: group),
                                              tls: NIOTSClientTLSProvider())
```

but for STARTTLS, we bundle `NIOTSConnectionBootstrap` with the TLS implementation from `swift-nio-ssl` (`NIOSSLClientTLSProvider`):

```
            bootstrap = try NIOClientTCPBootstrap(NIOTSConnectionBootstrap(group: group),
                                                  tls: NIOSSLClientTLSProvider(context: sslContext,
                                                                               serverHostname: hostname))
```

Reply:
@adam-fowler we need to do three things:
1. modify how we create bootstrap to something like this:
```
fileprivate static func makeBootstrap(on eventLoop: EventLoop, host: String, requiresTLS: Bool, configuration: HTTPClient.Configuration) throws -> NIOClientTCPBootstrap {
        let bootstrap: NIOClientTCPBootstrap
        #if canImport(Network)
        if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *), eventLoop is NIOTSEventLoop {
            if configuration.proxy != nil, requiresTLS {
                let tlsConfiguration = configuration.tlsConfiguration ?? TLSConfiguration.forClient()
                let sslContext = try NIOSSLContext(configuration: tlsConfiguration)
                let hostname = (!requiresTLS || host.isIPAddress) ? nil : host
                bootstrap = try NIOClientTCPBootstrap(NIOTSConnectionBootstrap(group: eventLoop), tls: NIOSSLClientTLSProvider(context: sslContext, serverHostname: hostname))
            } else {
                let tlsProvider = NIOTSClientTLSProvider(tlsOptions: .init())
                bootstrap = NIOClientTCPBootstrap(NIOTSConnectionBootstrap(group: eventLoop), tls: tlsProvider)
            }
        } else {
            bootstrap = try ClientBootstrap.makeBootstrap(on: eventLoop, host: host, requiresTLS: requiresTLS, configuration: configuration)
        }
        #else
        bootstrap = try ClientBootstrap.makeBootstrap(on: eventLoop, host: host, requiresTLS: requiresTLS, configuration: configuration)
        #endif

        if requiresTLS, configuration.proxy == nil {
            return bootstrap.enableTLS()
        }
        return bootstrap
    }
```
2. we need to revert `handshake` promise and add `addSSLHandlerIfNeeded` back
3. in `addSSLHandlerIfNeeded` we need to add `SSLHandler` if `eventLoopGroup` is not TS or if TLS & proxy enabled
If you have questions feel free to ping me on swiftpm/swift-server slack

Reply:
Sorry missed that case where the SSL handler should be applied after the proxy handler is processed. You changes all makes sense. I should be able to get them working.

Reply:
fixed